### PR TITLE
Signals

### PR DESCRIPTION
**Pull request by AndreaLanfranchi**
Introduce `SignalHandler` to enable (everywhere) check for process being signalled and allow neat shutdown

### CODE REVIEW

File: node/silkworm/common/signal_handler.cpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+*/
+
+#include "signal_handler.hpp"
+
+#include <csignal>
+#include <iostream>
+
+namespace silkworm {
+
+uint32_t SignalHandler::sig_count_{0};
```
Comment (yperbasis) [inline node/silkworm/common/signal_handler.cpp:24]: I'd use default initializers in the header, but it's a matter of preference.

File: node/silkworm/common/signal_handler.hpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+//! \brief Handler with static storage for signals sig
+class SignalHandler {
+  public:
+    static void init();                                           // Enable the hooks
+    static void handle(int sig_code);                             // Handles incoming signal
+    [[nodiscard]] static bool signalled() { return signalled_; }  // Whether a signal has been intercepted
+    static void reset();                                          // Reset to un-signalled (see tests coverage)
+
+  private:
+    static uint32_t sig_count_;
```
Comment (yperbasis) [inline node/silkworm/common/signal_handler.hpp:34]: Shouldn't these guys be atomic? Signal handling should be thread-safe (https://en.cppreference.com/w/cpp/utility/program/signal).

File: cmd/check_tx_lookup.cpp [hunk 3]
```diff
@@ -131,15 +117,15 @@ int main(int argc, char* argv[]) {
                 SILKWORM_LOG(LogLevel::Info) << "Scanned blocks " << expected_block_number << std::endl;
             }
 
-            if (should_stop_) {
+            if (SignalHandler::signalled()) {
```
Comment (greg7mdp) [inline cmd/check_tx_lookup.cpp:120]: should be `signaled` with one `l`

Comment (yperbasis) [reply cmd/check_tx_lookup.cpp:120]: signalled is also correct: https://www.merriam-webster.com/dictionary/signal

File: cmd/check_senders.cpp [hunk 4]
```diff
@@ -31,24 +28,14 @@
...
-#include <silkworm/types/block.hpp>
 
 using namespace silkworm;
 
 std::unique_ptr<stagedsync::recovery::RecoveryFarm> farm;
 
-void sig_handler(int) {
-    std::cout << std::endl << " Got interrupt. Stopping ..." << std::endl << std::endl;
-    if (farm) {
-        farm->stop();
```
Comment (greg7mdp) [inline cmd/check_senders.cpp:48]: Is the call to `farm->stop();` not needed anymore?

Comment (greg7mdp) [reply cmd/check_senders.cpp:48]: Nice change overall!

File: cmd/check_tx_lookup.cpp (hunk 3 above)
Comment (greg7mdp) [reply cmd/check_tx_lookup.cpp:120]: you are right, my bad. It is less common (about 10 times fewer results on google), but yes!

File: node/silkworm/common/signal_handler.cpp (hunk 1 above)
Comment (AndreaLanfranchi) [reply node/silkworm/common/signal_handler.cpp:24]: Right. (I had issues cause I forgot to inline the initializer)

File: node/silkworm/common/signal_handler.hpp (hunk 2 above)
Comment (AndreaLanfranchi) [reply node/silkworm/common/signal_handler.hpp:34]: Right.

File: cmd/check_senders.cpp (hunk 4 above)
Comment (AndreaLanfranchi) [reply cmd/check_senders.cpp:48]: Hmm. It was already removed in commit 2b0cd28
Maybe you were reviewing the initial commit ?

File: node/silkworm/common/signal_handler.cpp (hunk 1 above)
Comment (AndreaLanfranchi) [reply node/silkworm/common/signal_handler.cpp]: Solved vy c265f4f

File: cmd/check_senders.cpp (hunk 4 above)
Comment (greg7mdp) [reply cmd/check_senders.cpp:48]: yes, it was removed. But before when handling a signal we called `farm->stop();`, and we don't call it anymore in the new version. Just wondering if it is not needed anymore?

Comment (AndreaLanfranchi) [reply cmd/check_senders.cpp:48]: Oh I see your point. You refer to method `stop()` for the farm class.
Well this is still needed : should any worker return an error in the recovery phase then the whole farm is stopped. This is a programmatic stop, not a signalled one.
See https://github.com/torquem-ch/silkworm/blob/master/node/silkworm/stagedsync/recovery/recovery_farm.cpp#L231-L303

**Review by yperbasis — APPROVED** [review]

**Pull request merged**